diff --git a/Makefile.in b/Makefile.in
index 0e2cc95..93fc415 100755
--- a/Makefile.in
+++ b/Makefile.in
@@ -103,6 +103,7 @@ mkdir-init:
 CC = @CC@
 CXX = @CXX@
 LIBTOOL = @LIBTOOL@
+LIBS = @LIBS@
 
 INCLUDES = -I$(srcdir)/lib -I.
 LDFLAGS = @LDFLAGS@
@@ -143,7 +144,7 @@ xmlwf/xmlfile.o: xmlwf/xmlfile.c
 xmlwf/codepage.o: xmlwf/codepage.c
 xmlwf/@FILEMAP@.o: xmlwf/@FILEMAP@.c
 xmlwf/xmlwf: $(XMLWF_OBJS) $(LIBRARY)
-	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY)
+	$(LINK_EXE) $(XMLWF_OBJS) $(LIBRARY) $(LIBS)
 
 examples/elements.o: examples/elements.c
 examples/elements: examples/elements.o $(LIBRARY)
diff --git a/xmlwf/readfilemap.c b/xmlwf/readfilemap.c
index 088dda5..b9e8cb4 100755
--- a/xmlwf/readfilemap.c
+++ b/xmlwf/readfilemap.c
@@ -20,6 +20,10 @@
 #include <unistd.h>
 #endif
 
+#ifdef __EMSCRIPTEN__
+#include <unistd.h>
+#endif
+
 #ifndef S_ISREG
 #ifndef S_IFREG
 #define S_IFREG _S_IFREG
